nwastdlib/asyncio_cache.py

```diff
@@ -0,0 +1,193 @@
+
+import structlog
+from redis.asyncio import Redis as AIORedis
+
+logger = structlog.get_logger(__name__)
+
+
+@runtime_checkable
+class SerializerProtocol(Protocol):
+    """Abstract base class to specifies how to build yur own serializer."""
+
+    @staticmethod
+    def deserialize(data: Any) -> Any:
+        ...
+
+    @staticmethod
+    def serialize(data: Any) -> Any:
+        ...
+
+
+class DefaultSerializer:
+    """Implementation of the default protocol that uses pickle."""
+
+    @staticmethod
+    def deserialize(data: Any) -> Any:
+        return pickle.loads(data)  # noqa S403
+
+    @staticmethod
+    def serialize(data: bytes | bytearray | str) -> Any:
+        return pickle.dumps(data)  # noqa S403
+
+
+def _deserialize(data: Any, serializer: SerializerProtocol) -> Any:
+    try:
+        data = serializer.deserialize(data)
+    except Exception as e:
+        logger.error(
+            "Cache deserialization failed, returning None, so cache will be overwritten with a new value", error=e
+        )
+        return None
+    return data
+
+
+def get_hmac_checksum(secret: str, message: bytes | bytearray | str) -> str:
+    if isinstance(message, str):
+        message = message.encode()
+    h = hmac.new(secret.encode(), message, hashlib.sha512)
+    return h.hexdigest()
+
+
+async def set_cache_value(
+    pool: AIORedis, cache_key: str, value: Any, expiry_seconds: int, serializer: SerializerProtocol
+) -> None:
+    await pool.setex(cache_key, expiry_seconds, serializer.serialize(value))
+
+
+async def get_cache_value(pool: AIORedis, cache_key: str, serializer: SerializerProtocol) -> Any:
+    serialized_value = await pool.get(cache_key)
+    return _deserialize(serialized_value, serializer)
+
+
+async def set_signed_cache_value(
+    pool: AIORedis, secret: str, cache_key: str, value: Any, expiry_seconds: int, serializer: SerializerProtocol
+) -> None:
+    pickled_value = serializer.serialize(value)
+    checksum = get_hmac_checksum(secret, pickled_value)
+    pipeline = pool.pipeline()
+    pipeline.setex(cache_key, expiry_seconds, pickled_value).setex(
+        f"{cache_key}-checksum", expiry_seconds, checksum.encode()
+    )
+    result_set_cache_value, result_set_cache_checksum = await pipeline.execute()
+    if not result_set_cache_value or not result_set_cache_checksum:
+        logger.warning(
+            "Cache not set", cache_key=cache_key, value_ok=result_set_cache_value, checksum_ok=result_set_cache_checksum
+        )
+
+
+async def get_signed_cache_value(pool: AIORedis, secret: str, cache_key: str, serializer: SerializerProtocol) -> Any:
+    pipeline = pool.pipeline()
+    pipeline.get(cache_key).get(f"{cache_key}-checksum")
+    pickled_value, cache_checksum = await pipeline.execute()
+    if not pickled_value or not cache_checksum:
+        return None
+
+    cache_checksum = cache_checksum.decode("utf-8")
+    if (checksum := get_hmac_checksum(secret, pickled_value)) != cache_checksum:
```

Comment:
https://docs.python.org/3/library/hmac.html#hmac.HMAC.hexdigest

> Warning
> 
> When comparing the output of [hexdigest()](https://docs.python.org/3/library/hmac.html?highlight=hmac#hmac.HMAC.hexdigest) to an externally supplied digest during a verification routine, it is recommended to use the [compare_digest()](https://docs.python.org/3/library/hmac.html?highlight=hmac#hmac.compare_digest) function instead of the == operator to reduce the vulnerability to timing attacks.

Reply:
I guess it isn't that relevant here since we're not using this for authentication